File: backend/src/test/java/com/skewwhiffy/auraltester/notation/factory/ClefFactoryTest.java
package com.skewwhiffy.auraltester.notation.factory;

import com.skewwhiffy.auraltester.notation.model.clef.Clef;
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import com.skewwhiffy.auraltester.test.util.TestData;
import static org.assertj.core.api.Assertions.assertThat;

import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import static org.mockito.Mockito.when;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.function.Supplier;

@ExtendWith(MockitoExtension.class)
class ClefFactoryTest {
  @Mock
  private NoteFactory noteFactory;
  @InjectMocks
  private ClefFactory clefFactory;

  private AbsoluteNote highLedgerNote;
  private AbsoluteNote lowLedgerNote;

  @BeforeEach
  void setUp() {
    highLedgerNote = TestData.random().absoluteNote();
    lowLedgerNote = TestData.random().absoluteNote();
  }

  @Test
  void initializedTrebleClefCorrectly(){
    when(noteFactory.getAbsoluteNote("a")).thenReturn(highLedgerNote);

    val actual = clefFactory.getTreble();

    assertThat(actual.abc()).isEqualTo("treble");
    assertThat(actual.lowLedgerNote()).isEqualTo(AbsoluteNote.getMiddleC());
    assertThat(actual.highLedgerNote()).isEqualTo(highLedgerNote);
  }

  @Test
  void initializesAltoClefCorrectly() {
    testGeneric("alto", "D,", "B", () -> clefFactory.getAlto());
  }

  @Test
  void initializesTenorClefCorrectly() {
    testGeneric("tenor", "B,,", "G", () -> clefFactory.getTenor());
  }

  @Test
  void initializesBassClefCorrectly() {
    when(noteFactory.getAbsoluteNote("E,,")).thenReturn(lowLedgerNote);

    val actual = clefFactory.getBass();

    assertThat(actual.abc()).isEqualTo("bass");
    assertThat(actual.lowLedgerNote()).isEqualTo(lowLedgerNote);
    assertThat(actual.highLedgerNote()).isEqualTo(AbsoluteNote.getMiddleC());
  }

  private void testGeneric(
          String abc,
    String expectedLowLedgerNote,
    String expectedHighLedgerNote,
    Supplier<Clef> getClef
  ) {
    when(noteFactory.getAbsoluteNote(expectedLowLedgerNote)).thenReturn(lowLedgerNote);
    when(noteFactory.getAbsoluteNote(expectedHighLedgerNote)).thenReturn(highLedgerNote);

    val actual = getClef.get();

    assertThat(actual.abc()).isEqualTo(abc);
    assertThat(actual.lowLedgerNote()).isEqualTo(lowLedgerNote);
    assertThat(actual.highLedgerNote()).isEqualTo(highLedgerNote);
  }
}

File: backend/src/test/kotlin/com/skewwhiffy/auraltester/notation/factory/ClefFactoryTest.kt
package com.skewwhiffy.auraltester.notation.factory

import com.ninjasquad.springmockk.MockkBean
import com.skewwhiffy.auraltester.controller.ClefType
import com.skewwhiffy.auraltester.notation.model.clef.Clef
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import com.skewwhiffy.auraltester.test.util.TestData
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
internal class ClefFactoryTest {
    companion object {
        @JvmStatic
        fun clefTypeTestCases(): List<Arguments> {
            return listOf(
                Arguments.of(ClefType.TREBLE, ClefFactory::treble),
                Arguments.of(ClefType.ALTO, ClefFactory::alto),
                Arguments.of(ClefType.TENOR, ClefFactory::tenor),
                Arguments.of(ClefType.BASS, ClefFactory::bass)
            )
        }
    }

    @MockkBean
    private lateinit var noteFactory: NoteFactory

    @Autowired
    private lateinit var clefFactory: ClefFactory
    private lateinit var highLedgerNote: AbsoluteNote
    private lateinit var lowLedgerNote: AbsoluteNote

    @BeforeEach
    fun setUp() {
        highLedgerNote = TestData.random.absoluteNote
        lowLedgerNote = TestData.random.absoluteNote
        every { noteFactory.getAbsoluteNote(any()) } returns mockk<AbsoluteNote>()
    }

    @ParameterizedTest
    @MethodSource("clefTypeTestCases")
    fun getFromClefTypeWorks(clefType: ClefType, getExpected: (ClefFactory) -> Clef) {
        val expected = getExpected(clefFactory)

        val actual = clefFactory.get(clefType)

        assertThat(actual).isEqualTo(expected)
    }

    @Test
    fun initializedTrebleClefCorrectly() {
        every { noteFactory.getAbsoluteNote("a") } returns (highLedgerNote)
        val actual = clefFactory.treble
        assertThat(actual.abc).isEqualTo("treble")
        assertThat(actual.lowLedgerNote).isEqualTo(AbsoluteNote.middleC)
        assertThat(actual.highLedgerNote).isEqualTo(highLedgerNote)
    }

    @Test
    fun initializesAltoClefCorrectly() {
        testGeneric("alto", "D,", "B") { clefFactory.alto }
    }

    @Test
    fun initializesTenorClefCorrectly() {
        testGeneric("tenor", "B,,", "G") { clefFactory.tenor }
    }

    @Test
    fun initializesBassClefCorrectly() {
        every { noteFactory.getAbsoluteNote("E,,") } returns lowLedgerNote
        val actual = clefFactory.bass
        assertThat(actual.abc).isEqualTo("bass")
        assertThat(actual.lowLedgerNote).isEqualTo(lowLedgerNote)
        assertThat(actual.highLedgerNote).isEqualTo(AbsoluteNote.middleC)
    }

    private fun testGeneric(
        abc: String,
        expectedLowLedgerNote: String,
        expectedHighLedgerNote: String,
        getClef: () -> Clef
    ) {
        every { noteFactory.getAbsoluteNote(expectedLowLedgerNote) } returns lowLedgerNote
        every { noteFactory.getAbsoluteNote(expectedHighLedgerNote) } returns highLedgerNote
        val actual = getClef()
        assertThat(actual.abc).isEqualTo(abc)
        assertThat(actual.lowLedgerNote).isEqualTo(lowLedgerNote)
        assertThat(actual.highLedgerNote).isEqualTo(highLedgerNote)
    }

}